overlord/devicestate/handlers_install.go

```diff
@@ -630,6 +648,108 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
 	return nil
 }
 
+var seedOpen = seed.Open
+
+var maybeApplyPreseededData = func(st *state.State, ubuntuSeedDir, sysLabel, writableDir string) (preseeded bool, err error) {
+	preseedArtifact := filepath.Join(ubuntuSeedDir, "systems", sysLabel, "preseed.tgz")
+	if !osutil.FileExists(preseedArtifact) {
+		return false, nil
+	}
+
+	model, err := findModel(st)
+	if err != nil {
+		return false, fmt.Errorf("internal error: cannot find model: %v", err)
+	}
+
+	// find preseed assertion, this is going to fail if it doesn't match the model
+	as, err := assertstate.DB(st).Find(asserts.PreseedType, map[string]string{
+		"series":       model.Series(),
+		"brand-id":     model.BrandID(),
+		"model":        model.Model(),
+		"system-label": sysLabel,
+	})
+	if asserts.IsNotFound(err) {
+		return false, fmt.Errorf("preseed.tgz artifact is present but preseed assertion for brand %q, model %q and system label %q couldn't be found", model.BrandID(), model.Model(), sysLabel)
+	}
+	if err != nil {
+		return false, fmt.Errorf("internal error: cannot get preseed assertion: %v", err)
+	}
+
+	// TODO: consider a writer that feeds the file to stdin of tar and calculates the digest at the same time.
+	sha3_384, _, err := osutil.FileDigest(preseedArtifact, crypto.SHA3_384)
+	if err != nil {
+		return false, fmt.Errorf("cannot calculate preseed artifact digest: %v", err)
+	}
+
+	preseedAs := as.(*asserts.Preseed)
+	digest, err := base64.RawURLEncoding.DecodeString(preseedAs.ArtifactSHA3_384())
+	if err != nil {
+		return false, fmt.Errorf("cannot decode preseed artifact digest")
+	}
+	if !bytes.Equal(sha3_384, digest) {
+		return false, fmt.Errorf("invalid preseed artifact digest")
+	}
+
+	logger.Noticef("apply preseed data: %q, %q", writableDir, preseedArtifact)
+	cmd := exec.Command("tar", "--extract", "--preserve-permissions", "--preserve-order", "--gunzip", "--directory", writableDir, "-f", preseedArtifact)
+	if err := cmd.Run(); err != nil {
+		return false, err
+	}
+
+	logger.Noticef("copying snaps")
+
+	deviceSeed, err := seedOpen(dirs.SnapSeedDir, sysLabel)
+	if err != nil {
+		return false, err
+	}
+	tm := timings.New(nil)
+
+	// XXX: this is expensive, but we need snap revisions. Any way around this?
+	if err := deviceSeed.LoadAssertions(nil, nil); err != nil {
+		return false, err
+	}
+	if err := deviceSeed.LoadMeta("", tm); err != nil {
+		return false, err
+	}
+
```

Comment:
we should double check that the (snap, revision) pairs we are processing match completely the one listed in preseed

Reply:
Indeed; I left this aspect out for the new handler code as I'm not sure how much will it be affected by the new approach.

Reply:
Done.